### ISSUE TITLE
Allometric integration failure

### ISSUE THREAD

**Issue by jkshuman** [NGEET/fates#612]
Testing a boreal species parameter file with @adrifoster and getting a failure in the allometric integration that appears to be associated with sapwood allometry. At the moment we are using the default allom_la_per_sa_int and slope values (0.8=int, 0=slope) for evergreen and deciduous boreal trees. Both simulations fail at 22 years month 6 of simulation with tree growth and biomass accumulation. Feedback and thoughts would be appreciated. tagging @rgknox @jenniferholm @glemieux 
The first error message in the lnd.log happens after is associated with this section of parteh code: https://github.com/jkshuman/fates/blob/17aaa879bc4660d5a673513f496efed2bf099797/parteh/PRTAllometricCarbonMod.F90#L765-L778

lnd.log error:
Plant Growth Integrator could not find
 a solution in less than          300  tries
 Aborting
 carbon_balance   4269.93214339732    
 deltaC  5.090155772444395E-004
 totalC   136.145378414415    
 leaf: T   43428.4191181343       0.000000000000000E+000
 fnrt: T   43428.4191181343       0.000000000000000E+000
 sap: T   1037547.56645889       0.000000000000000E+000
 store: T   52114.1029417612       0.000000000000000E+000
 dead:   54607.7666557312       0.000000000000000E+000
 ENDRUN:
 ERROR in PRTAllometricCarbonMod.F90 at line 765    

Followed this failure with a simulation with increased integration steps set to 500 (up from 300) and including verbose logging in FatesAllometryMod set to "true" in the hopes that it will flag issues in the "CheckIntegratedAllometries"
https://github.com/jkshuman/fates/blob/17aaa879bc4660d5a673513f496efed2bf099797/biogeochem/FatesAllometryMod.F90#L173

With verbose logging and increased iteration steps, we still get a failure. The failure is associated with sapwood allometry:
Error from lnd.log (with verbose logging and steps =500)
FATES Dynamics:   22-07-13
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    244135.997830670     
 bsap (diagnosed):    244137.060407465     
 relative error:   4.352378097962951E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    419639.943573663     
 bsap (diagnosed):    419641.520727248     
 relative error:   3.758335406081040E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    476472.082156579     
 bsap (diagnosed):    476473.797062298     
 relative error:   3.599160603245344E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    228011.005324102     
 bsap (diagnosed):    228012.020169540     
 relative error:   4.450841837125617E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    628898.809591638     
 bsap (diagnosed):    628900.921529907     
 relative error:   3.358141476174558E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    628898.838311441     
 bsap (diagnosed):    628899.893314340     
 relative error:   1.677537093329176E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    594619.067965340     
 bsap (diagnosed):    594621.096332297     
 relative error:   3.411192386834351E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    594619.095741683     
 bsap (diagnosed):    594620.108990048     
 relative error:   1.704026402046414E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    715155.914350249     
 bsap (diagnosed):    715158.224250005     
 relative error:   3.229914273039127E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    715155.945136756     
 bsap (diagnosed):    715157.099049607     
 relative error:   1.613509608316129E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    769227.976928295     
 bsap (diagnosed):    769230.418991343     
 relative error:   3.174683407738909E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    769228.009258953     
 bsap (diagnosed):    769229.229203325     
 relative error:   1.585930858484564E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    424030.824788569     
 bsap (diagnosed):    424032.412782920     
 relative error:   3.744983409710965E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040842.51228536     
 bsap (diagnosed):    1040845.54641446     
 relative error:   2.915061807335038E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040842.55083697     
 bsap (diagnosed):    1040844.06660574     
 relative error:   1.456288043692690E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040842.56046002     
 bsap (diagnosed):    1040843.69741696     
 relative error:   1.092341672433292E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040842.56286393     
 bsap (diagnosed):    1040843.60521492     
 relative error:   1.001448230567157E-006
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040636.62032311     
 bsap (diagnosed):    1040637.62721867     
 relative error:   9.675755910010793E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040636.62047330     
 bsap (diagnosed):    1040637.62145940     
 relative error:   9.618968961310686E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040688.10629623     
 bsap (diagnosed):    1040689.10654365     
 relative error:   9.611395085682981E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040713.84921708     
 bsap (diagnosed):    1040714.84927986     
 relative error:   9.609383198486486E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040726.72067985     
 bsap (diagnosed):    1040727.72069647     
 relative error:   9.608820853982168E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040733.15641182     
 bsap (diagnosed):    1040734.15641690     
 relative error:   9.608650569152573E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040736.37427795     
 bsap (diagnosed):    1040737.37428015     
 relative error:   9.608593140442981E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040734.76534496     
 bsap (diagnosed):    1040735.76534500     
 relative error:   9.608587206002695E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040736.37427801     
 bsap (diagnosed):    1040737.37427804     
 relative error:   9.608572353797350E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040737.17874454     
 bsap (diagnosed):    1040738.17874458     
 relative error:   9.608564924356021E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040736.97762791     
 bsap (diagnosed):    1040737.97762791     
 relative error:   9.608566453411084E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040737.17874455     
 bsap (diagnosed):    1040738.17874455     
 relative error:   9.608564592136843E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040737.22902370     
 bsap (diagnosed):    1040738.22902371     
 relative error:   9.608564121225477E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040737.20388413     
 bsap (diagnosed):    1040738.20388413     
 relative error:   9.608564349969655E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040737.19131434     
 bsap (diagnosed):    1040738.19131434     
 relative error:   9.608564466019621E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040737.18502944     
 bsap (diagnosed):    1040738.18502944     
 relative error:   9.608564521807441E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040737.43642524     
 bsap (diagnosed):    1040738.43642524     
 relative error:   9.608562201927167E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040737.43485401     
 bsap (diagnosed):    1040738.43485401     
 relative error:   9.608562215314824E-007
 disparity in integrated/diagnosed sapwood carbon
 resulting from the on-allometry growth integration step
 bsap (integrated):    1040737.43406840     
 bsap (diagnosed):    1040738.43406840     
 relative error:   9.608562223686527E-007
 Plant Growth Integrator could not find
 a solution in less than          500  tries
 Aborting
 carbon_balance   4269.93214339732     
 deltaC  5.090155772444395E-004
 totalC   136.043575298965     
 leaf: T   43428.4191181343       0.000000000000000E+000
 fnrt: T   43428.4191181343       0.000000000000000E+000
 sap: T   1037547.56645889       0.000000000000000E+000
 store: T   52114.1029417612       0.000000000000000E+000
 dead:   54607.7666557312       0.000000000000000E+000
 ENDRUN:
 ERROR in PRTAllometricCarbonMod.F90 at line 765

**Comment by adrifoster** [comment]
Actually mine fails at year 35, month 12 but for the same reason. 

I noticed that with our input parameters bsap calculated from bsap_ltarg_slatop is generally much higher (sometimes twice as high) than that calculated with bagw_allom + bbgw_allom. So the model is always capping it at bagw_allom + bbgw_allom.

**Comment by rgknox** [comment]
That plant reported in the last log line has 1037 metric tons of sapwood.  And the leaf, fineroot and storage are on par with deadwood.  
Looking through that log, it seems like the integrator is just flat out having a tough time predicting the forward sapwood pool.  This could be because the allometric curve for sapwood in this has a lot of curvature and it seems like sapwood biomass grows rapidly and non-linearly with DBH.  I'm guessing your runs are taking a long time because it is doing so many re-tries.
The relative error on the solve is quite low though (because the absolute is so high I suppose).  

The log shows several cohorts in it, but you can kind of tell where it just starts reporting out on the cohort that finally kills the integrator.  I see that it fails a substep on the order of some multiplier of 10s of times, but I think what has happened here, is that the delta-carbon has decreased so much, that even if most of the steps are successful, it still can't complete the integration in 300 tries.

TLDR: My take:  Since the algorithm is self correcting, ie if it way overestimates one pool, it will correct in the next time-steps, perhaps we can bump up the maximum error tolerance to a larger value.  Perhaps it would be better to incorporate a relative error to work along with an absolute error, like it has to be less than 10kg and 1% of the total mass.  Also, the allometry seems really weird, is this a real tree.  1000Mg of sapwood?  Different integrator (rk45)?

**Comment by jkshuman** [comment]
@rgknox thanks for the comments. sounds like we need to look at our suite of allometry values.

**Comment by jkshuman** [comment]
@adrifoster and I looked into this and it appears this failure was due to high SLA that resulted in unreasonable sapwood values. closing.
